src/main/java/bio/terra/workspace/service/workspace/flight/GoogleCloudSyncStep.java

```diff
@@ -0,0 +1,103 @@
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.exception.RetryException;
+import bio.terra.workspace.service.iam.model.IamRole;
+import bio.terra.workspace.service.workspace.CloudSyncRoleMapping;
+import com.google.api.services.cloudresourcemanager.model.Binding;
+import com.google.api.services.cloudresourcemanager.model.GetIamPolicyRequest;
+import com.google.api.services.cloudresourcemanager.model.Policy;
+import com.google.api.services.cloudresourcemanager.model.SetIamPolicyRequest;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+
+/**
+ * A {@link Step} that grants GCP IAM permissions to Sam policy groups.
+ *
+ * <p>This step will grant GCP IAM roles to the google groups underlying Sam policies. It follows a
+ * read-modify-write pattern using GCP's field eTag to ensure the write does not clobber other
+ * changes. The read-modify-write pattern is necessary to support existing IAM groups which the
+ * Buffer Service may grant on projects before handing them out.
+ *
+ * <p>The "modify" part of this step specifically adds GCP bindings as specified in {@link
+ * CloudSyncRoleMapping}. Note that the bindings list sent to GCP may contain multiple entries with
+ * the same role. This is valid, though GCP will condense them into one binding per role internally.
+ */
+public class GoogleCloudSyncStep implements Step {
+
+  CloudResourceManagerCow resourceManagerCow;
+
+  private Logger logger = LoggerFactory.getLogger(GoogleCloudSyncStep.class);
+
+  @Autowired
+  public GoogleCloudSyncStep(CloudResourceManagerCow resourceManagerCow) {
+    this.resourceManagerCow = resourceManagerCow;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+    String gcpProjectName = flightContext.getWorkingMap().get(GOOGLE_PROJECT_ID, String.class);
+    FlightMap workingMap = flightContext.getWorkingMap();
+    try {
+      Policy currentPolicy =
+          resourceManagerCow
+              .projects()
+              .getIamPolicy(gcpProjectName, new GetIamPolicyRequest())
+              .execute();
+      List<Binding> existingBindings = currentPolicy.getBindings();
+
+      Map<IamRole, String> samEmailMap =
+          Map.of(
+              IamRole.READER,
+                  workingMap.get(WorkspaceFlightMapKeys.IAM_READER_GROUP_EMAIL, String.class),
+              IamRole.WRITER,
+                  workingMap.get(WorkspaceFlightMapKeys.IAM_WRITER_GROUP_EMAIL, String.class),
+              IamRole.OWNER,
+                  workingMap.get(WorkspaceFlightMapKeys.IAM_OWNER_GROUP_EMAIL, String.class));
+      List<Binding> newBindings = new ArrayList<>();
+      for (IamRole iamRole : IamRole.values()) {
```

Comment:
This code assumes that the list of IamRole.values() is identical to the roles in the map. Might there be more roles than there are sync'd roles?

Also, it doesn't make sense to me to build the map and then take it apart to build the bindings.Seems like it'd be easier to directly build the bindings.  Maybe a method to build a binding for a role. What am I missing?

Reply:
Looking at this again, that was mostly trying to be clever and avoid copying code. I think just building the bindings ends up being clearer and avoiding the potential mismatch from reading `IamRole.values()`.